homeassistant/components/mqtt/__init__.py

```diff
@@ -370,18 +389,30 @@ def async_publish_service(call):
         msg_topic = call.data[ATTR_TOPIC]
         payload = call.data.get(ATTR_PAYLOAD)
         payload_template = call.data.get(ATTR_PAYLOAD_TEMPLATE)
+        payload_file_path = call.data.get(ATTR_PAYLOAD_FILE_PATH)
         qos = call.data[ATTR_QOS]
         retain = call.data[ATTR_RETAIN]
-        try:
-            if payload_template is not None:
+        if payload_template is not None:
+            try:
                 payload = \
                     template.Template(payload_template, hass).async_render()
-        except template.jinja2.TemplateError as exc:
-            _LOGGER.error(
-                "Unable to publish to '%s': rendering payload template of "
-                "'%s' failed because %s",
-                msg_topic, payload_template, exc)
-            return
+            except template.jinja2.TemplateError as exc:
+                _LOGGER.error(
+                    "Unable to publish to '%s': rendering payload template of "
+                    "'%s' failed because %s",
+                    msg_topic, payload_template, exc)
+                return
+        elif payload_file_path is not None:
```

Comment:
Paulus speak about recevie binary payloads. And that is supported with you change. So sending binary data is a other Topic.

First a service is a technic from hass and that is limited. So you can not send binary data trought a service call. It is also not the idea to use base64 for that. It is also not the idea that the mqtt service have full mqtt compatibility while it is a Service and give support to use it with scripts and automation they also not support binarys.

For binary stuff und IPC we have make dispatcher they allow to send data trought hass without passing trought the event bus. 

So we have 2 variants to do binary stuff on our components:
- the component access to hass.data[mqtt] object and send his data (my favor)
- we make a general dispatcher on mqtt component they support binary

I think you PR should first support to receive binary. For sending stuff, do that on your i.e. notify or what ever component.